internal/ethapi/api.go

```diff
@@ -706,64 +706,68 @@ func (s *PublicBlockChainAPI) GetHeaderByHash(ctx context.Context, hash common.H
 //     only the transaction hash is returned.
 func (s *PublicBlockChainAPI) GetBlockByNumber(ctx context.Context, number rpc.BlockNumber, fullTx bool) (map[string]interface{}, error) {
 	block, err := s.b.BlockByNumber(ctx, number)
-	if block != nil && err == nil {
-		response, err := s.rpcMarshalBlock(ctx, block, true, fullTx)
-
-		if err == nil {
+	if block == nil || err != nil {
+		return nil, err
+	}
+	response, err := s.rpcMarshalBlock(ctx, block, true, fullTx)
+	if err == nil {
+		if s.b.RPCEthCompatibility() {
 			addEthCompatibilityFields(ctx, response, s.b, block)
-			if number == rpc.PendingBlockNumber {
-				// Pending blocks need to nil out a few fields
-				for _, field := range []string{"hash", "nonce", "miner"} {
-					response[field] = nil
-				}
+		}
+		if number == rpc.PendingBlockNumber {
+			// Pending blocks need to nil out a few fields
+			for _, field := range []string{"hash", "nonce", "miner"} {
+				response[field] = nil
 			}
 		}
-		return response, err
 	}
-	return nil, err
+	return response, err
 }
 
 // GetBlockByHash returns the requested block. When fullTx is true all transactions in the block are returned in full
 // detail, otherwise only the transaction hash is returned.
 func (s *PublicBlockChainAPI) GetBlockByHash(ctx context.Context, hash common.Hash, fullTx bool) (map[string]interface{}, error) {
 	block, err := s.b.BlockByHash(ctx, hash)
-	if block != nil {
-		result, err := s.rpcMarshalBlock(ctx, block, true, fullTx)
-		if err != nil {
-			return nil, err
-		}
+	if block == nil {
+		return nil, err
+	}
+	result, err := s.rpcMarshalBlock(ctx, block, true, fullTx)
+	if err != nil {
+		return nil, err
+	}
+	if s.b.RPCEthCompatibility() {
 		addEthCompatibilityFields(ctx, result, s.b, block)
-		return result, nil
 	}
-	return nil, err
+	return result, nil
 }
 
 // addEthCompatibilityFields seeks to work around the incompatibility of celo
 // and ethers.js (and potentially other web3 clients) by adding fields to our
 // rpc response that ethers.js depends upon.
 // See https://github.com/celo-org/celo-blockchain/issues/1945
 func addEthCompatibilityFields(ctx context.Context, response map[string]interface{}, b Backend, block *types.Block) {
-	isGingerbread := b.ChainConfig().IsGingerbread(block.Number())
-	if !b.RPCEthCompatibility() {
-		if !isGingerbread {
-			delete(response, "gasLimit")
-		}
-		return
-	}
-
```

Comment:
Why this change? Isn't it better to check inside `addEthCompatibilityFields` rather than at every call site?

Reply:
If you are referring to the check for rpc compatibility, I prefer to have the check outside rather than inside, so that from the top level code reads if X then do Y rather than just do Y (but then internally, oh actually don't do anything at all if !X).

If you are referring to the removal of the deletion of the gasLimit I updated RPCMarshalHeader to only add the gas limit if it the rlp decoded block had a gas limit set and RPCMarshalHeader is always called regardless of the eth compatibility setting.

Reply:
> If you are referring to the check for rpc compatibility, I prefer to have the check outside rather than inside, so that from the top level code reads if X then do Y rather than just do Y (but then internally, oh actually don't do anything at all if !X).

I prefer not to repeat the checking code and avoiding the possibility to forget the check. But I can live with either approach.